sources/iTermPowerManager.m

```diff
@@ -11,29 +11,52 @@
 
 NSString *const iTermPowerManagerStateDidChange = @"iTermPowerManagerStateDidChange";
 
-@implementation iTermPowerManager
+@implementation iTermPowerManager {
+    CFRunLoopRef _runLoop;
+    CFRunLoopSourceRef _runLoopSource;
+}
+
+static void ITPMpowerSourceChanged(void *context) {
+    iTermPowerManager* pm = (__bridge iTermPowerManager *)(context);
+    
+    CFStringRef source = IOPSGetProvidingPowerSourceType(NULL);
+    if (source){
+        pm.connectedToPower = [@"AC Power" isEqualToString:(__bridge NSString *)(source)];
+    }
+}
 
 + (instancetype)sharedInstance {
-    return [[self alloc] initPrivate];
+    static iTermPowerManager *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initPrivate];
+    });
+    
+    return sharedInstance;
 }
 
 - (instancetype)initPrivate {
     self = [super init];
     if (self) {
-        [self updateCharging];
-        [NSTimer it_scheduledTimerWithTimeInterval:5 repeats:YES block:^(NSTimer * _Nonnull timer) {
-            BOOL before = _connectedToPower;
-            [self updateCharging];
-            if (_connectedToPower != before) {
-                [[NSNotificationCenter defaultCenter] postNotificationName:iTermPowerManagerStateDidChange object:nil];
-            }
-        }];
+        _runLoop = CFRunLoopGetMain();
+        _runLoopSource = IOPSCreateLimitedPowerNotification(ITPMpowerSourceChanged,(__bridge void *)(self));
+        if (_runLoop && _runLoopSource){
+            CFRunLoopAddSource(_runLoop,_runLoopSource,kCFRunLoopDefaultMode);
+            ITPMpowerSourceChanged((__bridge void *)(self));
+        }
     }
     return self;
 }
 
-- (void)updateCharging {
-    _connectedToPower = (IOPSGetTimeRemainingEstimate() == kIOPSTimeRemainingUnlimited);
+- (void)setConnectedToPower:(BOOL)connectedToPower {
+    if (_connectedToPower != connectedToPower) {
+        _connectedToPower = connectedToPower;
```

Comment:
Is the notification delivered not on the main queue, necessitating the dispatch_async? If so the assignment to `_connectedToPower` is unsafe and needs to be protected by a lock of some sort. I'd use `@synchronized(self)` to keep it simple as this will happen very rarely and performance isn't a concern.

Reply:
The crash happens when the notification is sent from the dispatch_once block.
```    0x103456138 <+80>:  ja     0x103456107               ; <+31>
    0x10345613a <+82>:  leaq   0x1fdc7(%rip), %rcx       ; "BUG IN CLIENT OF LIBDISPATCH: trying to lock recursively"
    0x103456141 <+89>:  movq   %rcx, 0x4ae08(%rip)       ; gCRAnnotations + 8
```
The call is from the main thread, my fix is just a workaround. I'll try to find what goes wrong.